src/main/java/org/springframework/data/couchbase/core/query/Query.java

```diff
@@ -110,30 +157,87 @@ public void appendSort(final StringBuilder sb) {
 		sb.append(" ORDER BY ");
 		sort.stream().forEach(order -> {
 			if (order.isIgnoreCase()) {
-				throw new IllegalArgumentException(String.format("Given sort contained an Order for %s with ignore case! "
-						+ "Couchbase N1QL does not support sorting ignoring case currently!", order.getProperty()));
+				throw new IllegalArgumentException(String.format(
+						"Given sort contained an Order for %s with ignore case! "
+								+ "Couchbase N1QL does not support sorting ignoring case currently!",
+						order.getProperty()));
 			}
 			sb.append(order.getProperty()).append(" ").append(order.isAscending() ? "ASC," : "DESC,");
 		});
 		sb.deleteCharAt(sb.length() - 1);
 	}
 
-	public void appendWhere(final StringBuilder sb) {
-		sb.append(" WHERE ");
-		boolean first = true;
-		for (QueryCriteria c : criteria) {
-			if (first) {
-				first = false;
-			} else {
-				sb.append(" AND ");
+	public void appendWhere(final StringBuilder sb, int[] paramIndexPtr) {
+		if (!criteria.isEmpty()) {
+			appendWhereOrAnd(sb);
+			boolean first = true;
+			for (QueryCriteria c : criteria) {
+				if (first) {
+					first = false;
+				} else {
+					sb.append(" AND ");
+				}
+				sb.append(c.export(paramIndexPtr));
+			}
+		}
+	}
+
+	public void appendCriteria(StringBuilder sb, QueryCriteria criteria){
+		appendWhereOrAnd(sb);
+		sb.append(criteria.export());
+	}
+
+	private void appendWhereOrAnd(StringBuilder sb) {
+
+		String querySoFar = sb.toString().toUpperCase();
+
+		Pattern WHERE_PATTERN = Pattern.compile("\\sWHERE\\s");
```

Comment:
compile the pattern once in the class and then reuse, since the compilation is slow

Reply:
fixed